### DISCUSSION TITLE
Chai v5 - Unknown file extension ".ts" when running with mocha v10

### DISCUSSION THREAD

**Discussion by matteobruni** [chaijs/chai#1594] [Support]
Hello,

I've already checked [this](https://github.com/chaijs/chai/discussions/1575) without success. I'd prefer not using the solution suggested by the author of that discussion, also I didn't understand well how to achieve it, but since it doesn't seem the right way, I'd prefer using a cleaner solution.

I've tried also doing changes using [this gist](https://gist.github.com/43081j/78ce1392abb5043b02a29355006880a5) but when I run `pnpm run test` or `mocha`, `mocha` fails in an unexpected way, like the command is called without some arguments.

You can see my tests here: https://github.com/tsparticles/tsparticles/tree/v3/utils/tests (updated to v5 here https://github.com/tsparticles/tsparticles/tree/chai-v5/utils/tests)
And the failing error here: https://github.com/tsparticles/tsparticles/pull/5277 in the failing GitHub Action

<img width="815" alt="Screenshot 2024-01-30 alle 13 00 12" src="https://github.com/chaijs/chai/assets/176620/ba2063c3-1526-426f-b2f9-9813c3c92488">

Can somebody help me fixing that?

For building the project (if needed):

First clone `v3` branch, it's more updated than `main`

```
pnpm i
pnpm run slimbuild
```

once done this the first time, only `utils/tests` can be built using `pnpm run build` (it executes also the tests)

The `chai` dependency is in the root `package.json` config.

**Comment by 43081j** [comment]
im gonna guess its this:
https://github.com/tsparticles/tsparticles/blob/69556009df1fa134c176da1f0eea83209fc2d3b8/utils/tests/.mocharc.json#L8

which you can see is different in the gist:
https://gist.github.com/43081j/78ce1392abb5043b02a29355006880a5#file-mocharc-json-L2

afaik you can specify both if somehow you have a mix of cjs/esm sources too (the `require` line _and_ the `loader` line)

**Reply by matteobruni** [reply]
I've created a branch for this, so it's easier to track changes and see the errors: https://github.com/tsparticles/tsparticles/pull/5296

There you can see the changes made, but it's returning errors about register.

✖ ERROR: Error: Cannot find module '~/tsparticles/tsparticles/node_modules/jsdom-global/register' imported from ~/tsparticles/tsparticles/node_modules/.pnpm/mocha@10.2.0/node_modules/mocha/lib/nodejs/esm-utils.js

I've tried also removing all the registers but the output is what I was describing in the post:

```
> nyc mocha

(node:56031) ExperimentalWarning: `--experimental-loader` may be removed in the future; instead use `register()`:
--import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("ts-node/esm", pathToFileURL("./"));'
(Use `node --trace-warnings ...` to show where the warning was created)
mocha inspect [spec..]

Run tests with Mocha

Rules & Behavior
      --allow-uncaught       Allow uncaught errors to propagate        [boolean]
  -A, --async-only           Require all tests to use a callback (async) or
                             return a Promise                          [boolean]
  -b, --bail                 Abort ("bail") after first test failure   [boolean]
      --check-leaks          Check for global variable leaks           [boolean]
      --delay                Delay initial execution of root suite     [boolean]
      --dry-run              Report tests without executing them       [boolean]
      --exit                 Force Mocha to quit after tests complete  [boolean]
      --fail-zero            Fail test run if no test(s) encountered   [boolean]
      --forbid-only          Fail if exclusive test(s) encountered     [boolean]
      --forbid-pending       Fail if pending test(s) encountered       [boolean]
      --global, --globals    List of allowed global variables            [array]
  -j, --jobs                 Number of concurrent jobs for --parallel; use 1 to
                             run in serial
                                   [number] [default: (number of CPU cores - 1)]
  -p, --parallel             Run tests in parallel                     [boolean]
      --retries              Retry failed tests this many times         [number]
  -s, --slow                 Specify "slow" test threshold (in milliseconds)
                                                          [string] [default: 75]
  -t, --timeout, --timeouts  Specify test timeout threshold (in milliseconds)
                                                        [string] [default: 2000]
  -u, --ui                   Specify user interface    [string] [default: "bdd"]

Reporting & Output
  -c, --color, --colors                     Force-enable color output  [boolean]
      --diff                                Show diff on failure
                                                       [boolean] [default: true]
      --full-trace                          Display full stack traces  [boolean]
      --inline-diffs                        Display actual/expected differences
                                            inline within each string  [boolean]
  -R, --reporter                            Specify reporter to use
                                                      [string] [default: "spec"]
  -O, --reporter-option,                    Reporter-specific options
  --reporter-options                        (<k=v,[k1=v1,..]>)           [array]

Configuration
      --config       Path to config file   [string] [default: (nearest rc file)]
  -n, --node-option  Node or V8 option (no leading "--")                 [array]
      --package      Path to package.json for config                    [string]

File Handling
      --extension          File extension(s) to load
                                           [array] [default: ["js","cjs","mjs"]]
      --file               Specify file(s) to be loaded prior to root suite
                           execution                   [array] [default: (none)]
      --ignore, --exclude  Ignore file(s) or glob pattern(s)
                                                       [array] [default: (none)]
      --recursive          Look for tests in subdirectories            [boolean]
  -r, --require            Require module              [array] [default: (none)]
  -S, --sort               Sort test files                             [boolean]
  -w, --watch              Watch files in the current working directory for
                           changes                                     [boolean]
      --watch-files        List of paths or globs to watch               [array]
      --watch-ignore       List of paths or globs to exclude from watching
                                      [array] [default: ["node_modules",".git"]]

Test Filters
  -f, --fgrep   Only run tests containing this string                   [string]
  -g, --grep    Only run tests matching this string or regexp           [string]
  -i, --invert  Inverts --grep and --fgrep matches                     [boolean]

Positional Arguments
  spec  One or more files, directories, or globs to test
                                                     [array] [default: ["test"]]

Other Options
  -h, --help             Show usage information & exit                 [boolean]
  -V, --version          Show version number & exit                    [boolean]
      --list-interfaces  List built-in user interfaces & exit          [boolean]
      --list-reporters   List built-in reporters & exit                [boolean]

✖ ERROR: null
----------------|---------|----------|---------|---------|-------------------
File            | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------------|---------|----------|---------|---------|-------------------
All files       |       0 |        0 |       0 |       0 |                   
 src            |       0 |        0 |       0 |       0 |                   
  ColorUtils.ts |       0 |        0 |       0 |       0 | 18-282            
  Options.ts    |       0 |        0 |       0 |       0 | 17-643            
  Particle.ts   |       0 |        0 |       0 |       0 | 7-166             
  Particles.ts  |       0 |        0 |       0 |       0 | 7-258             
  QuadTree.ts   |       0 |        0 |       0 |       0 | 6-196             
  Themes.ts     |       0 |        0 |       0 |       0 | 5-134             
  Utils.ts      |       0 |        0 |       0 |       0 | 31-695            
 src/Fixture    |       0 |      100 |     100 |       0 |                   
  Window.ts     |       0 |      100 |     100 |       0 | 3                 
----------------|---------|----------|---------|---------|-------------------
 ELIFECYCLE  Test failed. See above for more details.
 ELIFECYCLE  Command failed with exit code 1.
```

**Reply by matteobruni** [reply]
Additional info: adding only the loader didn't work, same error about registers, so I tried adding `type module` in the `package.json` and `ES` build in the `tsconfig.json`

**Reply by 43081j** [reply]
this'll probably come down to the fact chai is esm only in 5.x

the gist uses `type: "module"` in its package.json. if you do the same in your repo, it'll be a breaking change keep in mind

if you stick to commonjs (by not having that `type` in your package.json), you'd have to use a dynamic import unfortunately (`const chai = await import('chai');`).

moving to ESM is no small task in most cases, so its fine to stay on chai 4.x for now (we're still doing security updates etc, feature set is the same roughly too)

if you want to move to ESM, you need:
- `type: "module"`
- tsconfig `module: "nodenext"` and `moduleResolution: "nodenext"`

**Reply by matteobruni** [reply]
I can migrate that test package to ESM without any issue, all the other packages are also exporting both ESM and CJS.

I've made both of these:

> * `type: "module"`
> * tsconfig `module: "nodenext"` and `moduleResolution: "nodenext"`

Also the tsconfig `target` is now `ESNext`

but that's where the last mocha command-like output comes out. Using `nyc` before `mocha` is not affecting anything, I've tried removing it but nothing changed.

I've noticed now that there were some non-existing folders referenced in the tsconfig ([commit](https://github.com/tsparticles/tsparticles/pull/5296/commits/548388132fcbb072326809f91bef9784305f5767)) but nothing changed again.

**Reply by 43081j** [reply]
that sounds like mocha is struggling to find the test files

maybe try explicitly defining the glob, like here:
https://gist.github.com/43081j/78ce1392abb5043b02a29355006880a5#file-mocharc-json-L4-L6

**Reply by matteobruni** [reply]
Unfortunately it didn't work ☹️

**Comment by glektarssza** [comment]
Might help you. I've been fighting with this all evening and I finally got my version of things to work.

https://github.com/glektarssza/gh-action-setup-odin/tree/b9261bb6ab4c769b125e08b451a1c0d88b5b89f7

The general gist of what I had to do was:

* Create my own `install-hooks.mjs` file.
* Load that file in via the `imports` configuration option for Mocha.
* Change ALL my TypeScript extensions to `.mts`.
* Set the `extension` option for Mocha to include `mts`.
* Stop using TypeScript path aliases.

Given the massive headache this has been I'm probably going to look into a different assertion framework instead of Chai. It's just not worth it to throw an entire project into chaos because one package decided to move to ESM and only ESM.

**Reply by 43081j** [reply]
i think you've come to the wrong conclusion there. you never _had_ to update to 5.x. if you're not in an ESM project, you can happily stick to 4.x until the day you move to ESM (if ever).

in an ESM project, it should be very easy and natural to use chai and mocha without much change

**Reply by glektarssza** [reply]
Are there plans to continue supporting 4.x? e.g. back porting changes or security fixes from 5.x+ to 4.x? If not then there will, eventually, be a need to migrate to 5.x+.

Personally I try to keep my dependencies up to date with the latest available version. This is very much a personal preference and I understand that the pain I experienced is of my own doing since there was, strictly, no need to upgrade.

It's worth noting, though, that I've run into this even in a fresh ESM-style project because of TypeScript. `ts-node` does not fully support ESM yet, particularly the `paths` feature of TypeScript. They recommend using another package, `tsconfig-paths`, for this and that package does not support ESM at all. I think that's the ultimate root of my particular struggles.

**Reply by 43081j** [reply]
Yes we are actively supporting 4.x. Many consumers are still on cjs projects so it wouldn't quite be fair to drop support

It is a shame ts-node has trouble with paths. Without paths, things do seem to work well in an esm project. Do you know if there's an issue in tsnode to solve that or a related discussion?

**Reply by glektarssza** [reply]
`ts-node` has the following ticket for their ESM loader: https://github.com/TypeStrong/ts-node/issues/1514
